bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/RowMerger.java

```diff
@@ -216,138 +213,99 @@ abstract void validateChunk(RowInProgress rowInProgess, ByteString previousKey,
+        chunk.getValue(), currentId.labels);
     }
 
-    public void completeMultiChunkCell() {
+    private final void completeMultiChunkCell() {
       Preconditions.checkArgument(hasChunkInProgess());
       ByteString value = ByteStringer.wrap(outputStream.toByteArray());
-      cellBuilderInProgress.withFamily(currentId.family).withQualifier(currentId.qualifier)
-          .withValue(value);
-      flatRowBuilder.addCell(cellBuilderInProgress.build());
+      flatRowBuilder.addCell(currentId.family, currentId.qualifier, currentId.timestampMicros,
+        value, currentId.labels);
       outputStream = null;
-      cellBuilderInProgress = null;
     }
 
     /**
      * update the current key with the new chunk info
      */
-    void updateCurrentKey(ReadRowsResponse.CellChunk chunk) {
-      if (currentId == null || isNewRowKey(chunk)) {
-        if (flatRowBuilder == null) {
-          flatRowBuilder = FlatRow.newBuilder().withRowKey(chunk.getRowKey());
-        }
+    private final void updateCurrentKey(ReadRowsResponse.CellChunk chunk) {
+      if (flatRowBuilder == null || isNewRowKey(chunk)) {
+        flatRowBuilder = FlatRow.newBuilder().withRowKey(chunk.getRowKey());
         currentId = new CellIdentifier(chunk);
       } else if (chunk.hasFamilyName()) {
-        currentId = currentId.nextKeyForFamily(chunk);
+        currentId.updateForFamily(chunk);
       } else if (chunk.hasQualifier()) {
-        currentId = currentId.nextKeyForQualifier(chunk);
+        currentId.updateForQualifier(chunk);
       } else {
-        currentId = currentId.nextKeyForTimestamp(chunk);
+        currentId.updateForTimestamp(chunk);
       }
     }
 
     private boolean isNewRowKey(ReadRowsResponse.CellChunk chunk) {
       ByteString rowKey = chunk.getRowKey();
-      return !rowKey.isEmpty() && !rowKey.equals(currentId.rowKey);
+      return !rowKey.isEmpty() && flatRowBuilder != null && !rowKey.equals(flatRowBuilder.getRowKey());
     }
 
-    public boolean hasChunkInProgess() {
+    private boolean hasChunkInProgess() {
       return outputStream != null;
     }
 
-    void addPartialCellChunk(ReadRowsResponse.CellChunk chunk) throws IOException {
+    private void addPartialCellChunk(ReadRowsResponse.CellChunk chunk) throws IOException {
       if (outputStream == null) {
         outputStream = new ByteArrayOutputStream(chunk.getValueSize());
-        cellBuilderInProgress = FlatRow.Cell.newBuilder()
-            .withTimestamp(chunk.getTimestampMicros())
-            .withLabels(chunk.getLabelsList());
       }
       chunk.getValue().writeTo(outputStream);
     }
 
-    public ByteString getRowKey() {
-      return currentId.rowKey;
+    private ByteString getRowKey() {
+      return flatRowBuilder == null ? null : flatRowBuilder.getRowKey();
+    }
+
+    private boolean hasRowKey() {
+      return flatRowBuilder != null;
+    }
+
+    private void reset() {
```

Comment:
I've seen reset() methods get out of sync with internal state and cause hard-to-diagnose problems. Is it worth the tradeoff? If so then maybe a comment along with the member variables that mentions keeping reset() up to date.